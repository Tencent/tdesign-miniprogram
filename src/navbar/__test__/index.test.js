import path from 'path';
import simulate from 'miniprogram-simulate';

describe('navbar', () => {
  const navbar = load(path.resolve(__dirname, `../navbar`), 't-navbar');

  it(`: style && customStyle`, async () => {
    const id = simulate.load({
      template: `<t-navbar class="navbar" style="{{style}}" customStyle="{{customStyle}}"></t-navbar>`,
      usingComponents: {
        't-navbar': navbar,
      },
      data: {
        style: 'color: red',
        customStyle: 'font-size: 9px',
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));
    const $navbar = comp.querySelector('.navbar >>> .t-navbar');
    // expect(comp.toJSON()).toMatchSnapshot();
    if (VIRTUAL_HOST) {
      expect($navbar.dom.getAttribute('style').includes(`${comp.data.style}; ${comp.data.customStyle}`)).toBeTruthy();
    } else {
      expect($navbar.dom.getAttribute('style').includes(`${comp.data.customStyle}`)).toBeTruthy();
    }
  });

  it(':base', async () => {
    const id = simulate.load({
      template: `<t-navbar visible="{{visible}}" animation="{{animation}}" title="test" />`,
      data: {
        visible: true,
        animation: true,
      },
      usingComponents: {
        't-navbar': navbar,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    expect(comp.toJSON()).toMatchSnapshot();

    // invisible
    comp.setData({ visible: false });
    await simulate.sleep();
    expect(comp.toJSON()).toMatchSnapshot();

    // without aniamtion
    comp.setData({ animation: false });
    await simulate.sleep();
    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(':fixed', async () => {
    const id = simulate.load({
      template: `<t-navbar visible fixed title="test" />`,
      usingComponents: {
        't-navbar': navbar,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    await simulate.sleep();

    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(':event', async () => {
    const handleBack = jest.fn();
    const handleSuccess = jest.fn();
    const handleFail = jest.fn();
    const handleComplete = jest.fn();
    const id = simulate.load({
      template: `<t-navbar
        id="base"
        visible
        left-arrow
        bind:go-back="handleBack"
        bind:success="handleSuccess"
        bind:fail="handleFail"
        bind:complete="handleComplete"
         />`,
      methods: {
        handleBack,
        handleSuccess,
        handleFail,
        handleComplete,
      },
      usingComponents: {
        't-navbar': navbar,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $base = comp.querySelector('#base');
    const $back = comp.querySelector('#base >>> .t-navbar__btn');

    // go back
    const mocked = jest.spyOn(wx, 'navigateBack');

    // mock success
    mocked.mockImplementation((option) => option.success());
    $back.dispatchEvent('tap');
    await simulate.sleep();
    expect(handleSuccess).toHaveBeenCalledTimes(1);
    expect(handleBack).toHaveBeenCalledTimes(1);

    // mock fail
    mocked.mockImplementation((option) => option.fail());
    $back.dispatchEvent('tap');
    await simulate.sleep();
    expect(handleFail).toHaveBeenCalledTimes(1);
    expect(handleBack).toHaveBeenCalledTimes(2);

    // mock complete
    mocked.mockImplementation((option) => option.complete());
    $back.dispatchEvent('tap');
    await simulate.sleep();
    expect(handleComplete).toHaveBeenCalledTimes(1);
    expect(handleBack).toHaveBeenCalledTimes(3);

    // mock delta < 1
    let hasBacked = false;
    $base.setData({ delta: 0 });
    mocked.mockImplementation(() => {
      hasBacked = true;
    });
    $back.dispatchEvent('tap');
    await simulate.sleep();
    expect(hasBacked).toBeFalsy();
  });

  it(':menu button', async () => {
    const id = simulate.load({
      template: `<t-navbar title="test" />`,
      usingComponents: {
        't-navbar': navbar,
      },
    });
    wx.getMenuButtonBoundingClientRect = () => ({
      top: 10,
      bottom: 20,
      left: 0,
      height: 10,
      width: 40,
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    await simulate.sleep();

    expect(comp.toJSON()).toMatchSnapshot();
    wx.getMenuButtonBoundingClientRect = undefined;
  });
});
