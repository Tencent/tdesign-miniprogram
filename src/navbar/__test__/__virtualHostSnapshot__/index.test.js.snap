// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navbar :base 1`] = `
<main>
  <wx-view
    class="t-navbar t-navbar--fixed t-navbar--visible-animation class t-class"
    style=""
  >
    <wx-view
      class="t-navbar__placeholder"
    />
    <wx-view
      class="t-navbar__content"
    >
      <wx-view
        class="t-navbar__left t-class-left"
      >
        <wx-view
          class="t-navbar__capsule t-class-capsule"
        />
      </wx-view>
      <wx-view
        class="t-navbar__center t-class-center"
      >
        <wx-text
          class="t-navbar__center-title t-class-title"
        >
          test
        </wx-text>
      </wx-view>
    </wx-view>
  </wx-view>
</main>
`;

exports[`navbar :base 2`] = `
<main>
  <wx-view
    class="t-navbar t-navbar--fixed t-navbar--hide-animation class t-class"
    style=""
  >
    <wx-view
      class="t-navbar__placeholder"
    />
    <wx-view
      class="t-navbar__content"
    >
      <wx-view
        class="t-navbar__left t-class-left"
      >
        <wx-view
          class="t-navbar__capsule t-class-capsule"
        />
      </wx-view>
      <wx-view
        class="t-navbar__center t-class-center"
      >
        <wx-text
          class="t-navbar__center-title t-class-title"
        >
          test
        </wx-text>
      </wx-view>
    </wx-view>
  </wx-view>
</main>
`;

exports[`navbar :base 3`] = `
<main>
  <wx-view
    class="t-navbar t-navbar--fixed t-navbar--hide class t-class"
    style=""
  >
    <wx-view
      class="t-navbar__placeholder"
    />
    <wx-view
      class="t-navbar__content"
    >
      <wx-view
        class="t-navbar__left t-class-left"
      >
        <wx-view
          class="t-navbar__capsule t-class-capsule"
        />
      </wx-view>
      <wx-view
        class="t-navbar__center t-class-center"
      >
        <wx-text
          class="t-navbar__center-title t-class-title"
        >
          test
        </wx-text>
      </wx-view>
    </wx-view>
  </wx-view>
</main>
`;

exports[`navbar :fixed 1`] = `
<main>
  <wx-view
    class="t-navbar t-navbar--fixed t-navbar--visible-animation class t-class"
    style=""
  >
    <wx-view
      class="t-navbar__placeholder"
    />
    <wx-view
      class="t-navbar__content"
    >
      <wx-view
        class="t-navbar__left t-class-left"
      >
        <wx-view
          class="t-navbar__capsule t-class-capsule"
        />
      </wx-view>
      <wx-view
        class="t-navbar__center t-class-center"
      >
        <wx-text
          class="t-navbar__center-title t-class-title"
        >
          test
        </wx-text>
      </wx-view>
    </wx-view>
  </wx-view>
</main>
`;

exports[`navbar :menu button 1`] = `
<main>
  <wx-view
    class="t-navbar t-navbar--fixed  class t-class"
    style="--td-navbar-padding-top:20px; --td-navbar-right:414px; --td-navbar-capsule-height: 10px; --td-navbar-capsule-width:40px;"
  >
    <wx-view
      class="t-navbar__placeholder"
    />
    <wx-view
      class="t-navbar__content"
    >
      <wx-view
        class="t-navbar__left t-class-left"
      >
        <wx-view
          class="t-navbar__capsule t-class-capsule"
        />
      </wx-view>
      <wx-view
        class="t-navbar__center t-class-center"
      >
        <wx-text
          class="t-navbar__center-title t-class-title"
        >
          test
        </wx-text>
      </wx-view>
    </wx-view>
  </wx-view>
</main>
`;
